docs: failover2 (#378)
aws/aws-advanced-nodejs-wrapper: docs/using-the-nodejs-wrapper/using-plugins/UsingTheFailover2Plugin.md

The Failover Plugin v2 uses an optimized approach where the process of detecting and confirming a cluster topology is delegated to a central topology monitoring component that runs a separate asynchronous monitoring task. When the topology is confirmed and a new writer is detected, each waiting connection can resume and reconnect to a required host. This design helps minimize resources required for failover processing and scales better compared to the Failover Plugin.

<div style="text-align:center"><img src="../../images/failover2.png"  alt="failover2 diagram"/></div>

If two connections encounter communication issues with their internal (physical) DB connections, each connection may send a request to the topology monitoring component (`MonitoringRdsHostListProvider` in _Picture 2_) for updated topology information reflecting the new writer. Both connections are notified as soon as the latest topology is available. Connection tasks can resume, continue with their suspended workflows, and reconnect to a reader or a writer host as needed.

> The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) updates topology periodically. Usually it uses a connection to a writer host to fetch a cluster topology. Using a connection to a writer host allows to get topology first hand without a risk of getting stale data as in case of fetching it from a reader. In some exceptional cases the monitoring component may (temporarily) use a reader connection to fetch a topology however it will switch back to a writer host as soon as possible.

Reviewer: ```suggestion
The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) updates topology periodically. Usually it uses a connection to a writer host to fetch a cluster topology. Using a connection to a writer host allows to get topology first hand without a risk of getting stale data as in case of fetching it from a reader. In some exceptional cases the monitoring component may temporarily use a reader connection to fetch a topology however it will switch back to a writer host as soon as possible.
```

<div style="text-align:center"><img src="../../images/failover2_panic.png"  alt="failover2 panic mode diagram"/></div>

When the cluster topology needs to be confirmed, the monitoring component opens new monitoring tasks, one for each host in the cluster. Each of these tasks tries to connect to a host and checks if the host is a writer. When Aurora failover occurs, the new writer host is the first host to reflect the true topology of the cluster. Other hosts connect to the new writer shortly after and update their local copies of the topology. Topology information acquired from a reader host may be outdated/inaccurate for a short period after failover. You can see a typical example of stale topology in the diagram above with `instance 3`. The stale topology incorrectly shows that `instance 3` is still a writer.

> The asynchronous tasks monitoring the topology stop when a new writer is detected. For 30 seconds after a new writer is detected (and after all waiting connections have been notified), topology continues to be updated at an increased rate. This allows time for all readers to appear in the topology, since 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.

Reviewer: ```suggestion
The asynchronous tasks monitoring the topology stop when a new writer is detected. For 30 seconds after a new writer is detected and after all waiting connections have been notified, topology continues to be updated at an increased rate. This allows time for all readers to appear in the topology, since 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.
```

The Failover Plugin v2 uses an optimized approach where the process of detecting and confirming a cluster topology is delegated to a central topology monitoring component that runs a separate asynchronous monitoring task. When the topology is confirmed and a new writer is detected, each waiting connection can resume and reconnect to a required host. This design helps minimize resources required for failover processing and scales better compared to the Failover Plugin.

<div style="text-align:center"><img src="../../images/failover2.png"  alt="failover2 diagram"/></div>
<i>Picture 2. Connections call MonitoringRdsHostListProvider, which is responsible for detecting the new writer.</i>
<br><br>

If two connections encounter communication issues with their internal (physical) DB connections, each connection may send a request to the topology monitoring component (`MonitoringRdsHostListProvider` in _Picture 2_) for updated topology information reflecting the new writer. Both connections are notified as soon as the latest topology is available. Connection tasks can resume, continue with their suspended workflows, and reconnect to a reader or a writer host as needed.

> The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) fetches and confirms topology periodically. Outside of failover it performs "slow rate monitoring" at intervals set by `clusterTopologyRefreshRateMs`. This ensures use of the most up to date topology. When failover is triggered, it enters "panic mode" and works to detect a writer. After a new writer has been detected, the `MonitoringRdsHostListProvider` enters "high rate monitoring", with intervals set by `clusterTopologyHighRefreshRateMs`. This allows time for all readers to appear in the topology. The default of 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.

Reviewer 2: check that `clusterTopologyHighRefreshRateMs` is the name in the wrapper properties

All improvements mentioned above help the Failover Plugin v2 to operate with improved performance and less demand for resources.

A summary of the key differences and between the `failover` and `failover2` plugins is outlined below.

With the `failover` plugin:

- Each connection performs its own failover process.
- Each connection fetches topology by calling the `RdsHostListProvider`.
- Topology may be fetched from a reader host and it may be stale.

With the `failover2` plugin:

- Each client delegates detection of the new writer to the `MonitoringRdsHostListProvider` and suspends until the new writer is confirmed.
- The `MonitoringRdsHostListProvider` tries to connect to every cluster host.
- The `MonitoringRdsHostListProvider` uses an "Am I a writer?" approach to avoid reliance on stale topology.
- The `MonitoringRdsHostListProvider` continues topology monitoring at an increased rate to ensure all cluster hosts appear in the topology.

## Using the Failover Plugin v2

> The Failover Plugin v2 will **NOT** be enabled by default. It must be explicitly included by adding the plugin code `failover2` to the [`plugins`](../UsingTheNodejsWrapper.md#plugins) parameter, or by adding it to the current [Configuration Profile](../UsingTheNodejsWrapper.md#configuration-profiles). After you load the plugin, the failover feature will be enabled.

Reviewer 2: ```suggestion
The Failover Plugin v2 will **NOT** be enabled by default. It must be explicitly included by adding the plugin code `failover2` to the [`plugins`](../UsingTheNodejsWrapper.md#plugins) parameter, or by adding it to the current [Configuration Profile](../UsingTheNodejsWrapper.md#configuration-profiles). After you load the plugin, the failover2 feature will be enabled.
```

The Failover Plugin v2 uses an optimized approach where the process of detecting and confirming a cluster topology is delegated to a central topology monitoring component that runs a separate asynchronous monitoring task. When the topology is confirmed and a new writer is detected, each waiting connection can resume and reconnect to a required host. This design helps minimize resources required for failover processing and scales better compared to the Failover Plugin.

<div style="text-align:center"><img src="../../images/failover2.png"  alt="failover2 diagram"/></div>
<i>Picture 2. Connections call MonitoringRdsHostListProvider, which is responsible for detecting the new writer.</i>
<br><br>

If two connections encounter communication issues with their internal (physical) DB connections, each connection may send a request to the topology monitoring component (`MonitoringRdsHostListProvider` in _Picture 2_) for updated topology information reflecting the new writer. Both connections are notified as soon as the latest topology is available. Connection tasks can resume, continue with their suspended workflows, and reconnect to a reader or a writer host as needed.

> The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) fetches and confirms topology periodically. Outside of failover it performs "slow rate monitoring" at intervals set by `clusterTopologyRefreshRateMs`. This ensures use of the most up to date topology. When failover is triggered, it enters "panic mode" and works to detect a writer. After a new writer has been detected, the `MonitoringRdsHostListProvider` enters "high rate monitoring", with intervals set by `clusterTopologyHighRefreshRateMs`. This allows time for all readers to appear in the topology. The default of 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.

Reviewer 3: ```suggestion
The topology monitoring component mentioned above (`MonitoringRdsHostListProvider`) fetches and confirms topology periodically. Outside of failover it performs "slow rate monitoring" at intervals set by `clusterTopologyRefreshRateMs`. This ensures use of the most up to date topology. When failover is triggered, it enters "panic mode" and will try to detect a writer. After a new writer has been detected, the `MonitoringRdsHostListProvider` enters "high rate monitoring", with intervals set by `clusterTopologyHighRefreshRateMs`. This allows time for all readers to appear in the topology. The default of 30 seconds is usually enough time for cluster failover to complete and cluster topology to stabilize.
```

| `enableClusterAwareFailover`         | Boolean |    No    | Set to `true` to enable the fast failover behavior offered by the AWS Advanced NodeJS Wrapper. Set to `false` for simple connections that do not require fast failover functionality.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `true`                                                                                                                                           |

> | `clusterTopologyRefreshRateMs`       | Number  |    No    | Cluster topology refresh rate in milliseconds when a cluster is not in failover. It refers to the regular, slow monitoring rate explained above.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `30000`                                                                                                                                          |

Reviewer 3: ```suggestion
| `clusterTopologyRefreshRateMs`       | Number  |    No    | Cluster topology refresh rate in milliseconds when a cluster is not in failover.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `30000`                                                                                                                                          |
```
If we can't link to the specific paragraph above, should we leave the reference out of the description? For the high refresh rate as well

> | `clusterTopologyHighRefreshRateMs`   | Number  |    No    | Interval of time in milliseconds to wait between attempts to update cluster topology after the writer has come back online following a failover event. It corresponds to the increased monitoring rate described earlier. Usually, the topology monitoring component uses this increased monitoring rate for 30s after a new writer was detected.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `100`                                                                                                                                            |

Reviewer 3: ```suggestion
| `clusterTopologyHighRefreshRateMs`   | Number  |    No    | Interval of time in milliseconds to wait between attempts to update cluster topology after the writer has come back online following a failover event. Usually, the topology monitoring component uses this increased monitoring rate for 30s after a new writer was detected.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `100`                                                                                                                                            |
```
